### PR TITLE
Fork Panel, Login Panel

### PR DESCRIPTION
**Pull request by balazsbajorics**
Malte made a nice "This is not your project, but you can fork it" panel. I implemented a Forking behavior to accompany it, plus wired in some complicated show-this/don't-show-this logic.

**Commit Details:**
- new UI called ForksGiven and ForkButton
- `useTriggerForkProject` and `triggerForkProject` that ask for a new project ID and fork the project using `save`
- moved `save` from dispatch.ts to persistence.ts
- added two new hooks `useGetProjectMetadata` and `useIsMyProject`

### CODE REVIEW

File: editor/src/components/common/server-hooks.ts [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+type ProjectMetadata = {
+  title: string | null
+  ownerName: string | null
+  ownerPicture: string | null
+}
+
+export function useGetProjectMetadata(projectId: string | null): ProjectMetadata {
+  const previousProjectIdRef = React.useRef<string | null>(null)
+  const [userData, setUserData] = React.useState<ProjectMetadata>({
```
Comment (seanparsons) [inline editor/src/components/common/server-hooks.ts:14]: I feel a little like this should be some kind of "unknown" value, as it's not that they're null, the first time we have no idea what the values are/could be.

File: editor/src/components/common/server-hooks.ts [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+      })
+    }
+  }
+
+  return userData
+}
+
+export function useIsMyProject(projectId: string | null): boolean {
+  const previousProjectIdRef = React.useRef<string | null>(null)
+  const [myProject, setMyProject] = React.useState<boolean>(false)
```
Comment (seanparsons) [inline editor/src/components/common/server-hooks.ts:46]: Similarly here, it surely isn't that it's not the user's project, the answer to that isn't currently known until some time later.

File: editor/src/components/navigator/left-pane.tsx [hunk 3]
```diff
@@ -92,17 +106,203 @@ export const LeftPaneComponent = betterReactMemo('LeftPaneComponent', () => {
           }
         }}
       >
-        {selectedTab === LeftMenuTab.Project ? <ProjectPane /> : null}
+        {!isMyProject ? <ForksGiven /> : null}
```
Comment (seanparsons) [inline editor/src/components/navigator/left-pane.tsx:109]: Double negatives. :(

File: editor/src/components/navigator/left-pane.tsx [hunk 4]
```diff
@@ -545,12 +746,21 @@ const ProjectPane = betterReactMemo('ProjectSettingsPanel', () => {
...
+      forkedFrom: store.editor.forkedFromProjectId,
     }
   }, 'ProjectSettingsPanel')
 
   const [name, changeProjectName] = React.useState(projectName)
   const [description, changeProjectDescription] = React.useState(projectDescription)
   const [requestingPreviewImage, setRequestingPreviewImage] = React.useState(false)
+  const { title: forkedFromTitle } = useGetProjectMetadata(forkedFrom)
+
+  const forkedFromText =
```
Comment (seanparsons) [inline editor/src/components/navigator/left-pane.tsx:758]: This looks like it's duplicated from earlier.

File: editor/src/components/navigator/left-pane.tsx [hunk 5]
```diff
@@ -648,108 +858,120 @@ const ProjectPane = betterReactMemo('ProjectSettingsPanel', () => {
...
+                >
+                  <Subdued>
+                    These help you organise your projects. We also use them when you embed or share
+                    your project on social media and chat apps.
+                  </Subdued>
+                </UIGridRow>
+
+                <UIGridRow padded variant='<---1fr--->|------172px-------|'>
+                  <span>Name</span>
+                  {userState.loginState.type !== 'LOGGED_IN' ? (
```
Comment (seanparsons) [inline editor/src/components/navigator/left-pane.tsx:887]: More double negatives. :)

File: editor/src/uuiui/widgets/avatar/avatar.tsx [hunk 6]
```diff
@@ -22,9 +23,9 @@ export class Avatar extends React.Component<AvatarProps> {
 
   render() {
     const imageURL =
-      User.isLoggedIn(this.props.loginState) && this.props.loginState.user.picture !== null
-        ? `url(${this.props.loginState.user.picture})`
-        : User.isLoggedIn(this.props.loginState)
+      this.props.userPicture != null
```
Comment (seanparsons) [inline editor/src/uuiui/widgets/avatar/avatar.tsx:26]: Nested ternaries with double negatives! Much as it was like this before...

**Review by seanparsons — APPROVED** [review]

File: editor/src/components/navigator/left-pane.tsx (hunk 3 above)
Comment (balazsbajorics) [reply editor/src/components/navigator/left-pane.tsx:109]: however, all of the lines above and below have the <Component> : null structure, I tried null : <ForksGiven /> here and that seemed worse

File: editor/src/components/navigator/left-pane.tsx (hunk 4 above)
Comment (balazsbajorics) [reply editor/src/components/navigator/left-pane.tsx:758]: I think it's fine. the copy might diverge, turning this into a component would be premature IMO

**Review by maltenuhn — APPROVED** [review]

**Pull request merged**